Sentry interceptor example

nediamond opened the pull request:
<!--- Note to EXTERNAL Contributors -->
<!-- Thanks for opening a PR! 
If it is a significant code change, please **make sure there is an open issue** for this. 
We work best with you when we have accepted the idea first before you code. -->

<!--- For ALL Contributors 👇 -->

## What was changed
<!-- Describe what has changed in this PR -->

added Sentry interceptor example based on discussion in [slack thread](https://temporalio.slack.com/archives/CTT84RS0P/p1666821089048929)

## Why?
<!-- Tell your future self why have you made these changes -->
Sentry is a popular error monitoring tool & interceptors aren't super intuitive to make from scratch, hopefully this can be a good starting point for folks

## Checklist
<!--- add/delete as needed --->

1. How was this tested:
Executed locally

2. Any docs updates needed?
not sure

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/temporalio/samples-python?pullRequest=23) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/temporalio/samples-python?pullRequest=23) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/temporalio/samples-python?pullRequest=23) it.</sub>

cretz commented on sentry/interceptor.py:
```diff
@@ -0,0 +1,75 @@
+            name=transaction_name
+        ):
+            sentry_sdk.set_tag("temporal.execution_type", "activity")
+            sentry_sdk.set_tag("temporal.activity.name", input.fn.__name__)
```
```suggestion
            sentry_sdk.set_tag("temporal.activity.name", temporalio.activity.info().activity_type)
```

There are also other tags from info you might want like `activity_id` (so it can be tied w/ the Temporal history/UI), `workflow_type`, `workflow_id`, `workflow_namespace`, `task_queue`, etc.

cretz commented on sentry/interceptor.py:
```diff
@@ -0,0 +1,75 @@
+
+class _SentryActivityInboundInterceptor(ActivityInboundInterceptor):
+    async def execute_activity(self, input: ExecuteActivityInput) -> Any:
+        transaction_name = input.fn.__module__ + "." + input.fn.__name__
```
```suggestion
        transaction_name = input.fn.__module__ + "." + input.fn.__qualname__
```

Sometimes activities are methods of classes. Then again, you could use `temporalio.activity.info().activity_type` which is the registered name with the worker.

cretz commented on sentry/interceptor.py:
```diff
@@ -0,0 +1,75 @@
+                scope.clear()
+
+
+class SentryInterceptor(Interceptor):
```
Can you put a docstring on here describing its purpose?

cretz commented on sentry/interceptor.py:
```diff
@@ -0,0 +1,75 @@
+            name=transaction_name
+        ):
+            sentry_sdk.set_tag("temporal.execution_type", "workflow")
+            sentry_sdk.set_tag("temporal.workflow.name", input.type.__name__)
```
```suggestion
            sentry_sdk.set_tag("temporal.workflow.name", temporalio.workflow.info().workflow_type)
```

Same here as activities. There is lots of stuff on `temporalio.workflow.info()` you might want in here

cretz commented on sentry/worker.py:
```diff
@@ -0,0 +1,69 @@
+    # Run a worker for the workflow
+    async with Worker(
+        client,
+        task_queue="hello-activity-task-queue",
```
```suggestion
        task_queue="sentry-task-queue",
```

cretz commented on sentry/worker.py:
```diff
@@ -0,0 +1,69 @@
+        result = await client.execute_workflow(
+            GreetingWorkflow.run,
+            "World",
+            id="hello-activity-workflow-id",
+            task_queue="hello-activity-task-queue",
```
```suggestion
            id="sentry-workflow-id",
            task_queue="sentry-task-queue",
```

cretz reviewed (commented):
Added a couple of comments. Looks great!

Some other that I think would help:

* Add a README.md in this folder w/ details on how to run (see how OTel sample does it in #18)
* Add a "dependency group" for `sentry` in `pyproject.toml` to add the `sentry_sdk` dependency (see how OTel sample does it in #18)
* Separate the `starter` from the `worker`. We are doing this in all of the non-hello samples, and we do it in all SDKs, to emphasize that workers are often completely separate from starters (see how OTel sample does it in #18)

If the Sentry SDK comes with utilities to capture/assert output in a test, a test in `tests/sentry` would be great. But not required or anything.

nediamond commented:
will try to implement the tests/readme/worker changes this weekend

cretz commented:
@nediamond - Wanted to follow up here. I think this is a great sample. Let me know if there's anything I can do to help move it along.

nediamond commented:
@cretz sorry got busy with life stuff will set some time aside for this before next week - feel clear on what needs to be done thanks

nediamond commented:
@cretz i reverted the `poetry.lock` file bc it looked like the checks didnt like it - could use your help generating that since it seems like it may be from a difference in my setup or something

added a super basic readme, but i think this is about as much as i can commit to for now

cretz commented:
> could use your help generating that since it seems like it may be from a difference in my setup or something

> added a super basic readme, but i think this is about as much as i can commit to for now

Ok, I will look to fixing that and improving your README

nathanielobrown commented:
Hey @nediamond and @cretz, I just wanted to share a pitfall I found in this approach for implementing Sentry via interceptors when using threaded activities. Because activities are run in a thread in a `ThreadPoolExecutor` or in a separate process but the interceptor is run in the main thread, the activity code gets a totally different Sentry scope/context than the interceptor. So any context generated _inside the activity_ (such as breadcrumbs, setting tags, context etc.) will **not** be included in the resulting Sentry issue. I also was experiencing some "leakage" of contexts, where transaction names would be wrong (I don't 100% understand what was happening there). I haven't tested with `async` activities, but I feel like _maybe_ they might have the same issue too because Sentry uses `contextvars`?

My solution was to add Sentry handling to a custom decorator that wraps the `@activity.defn` decorator and does Sentry handling in the thread. I left the activity interceptor as a fallback, and set an attribute on the activity function and skip interceptor Sentry handling if this attribute is set.

I just wanted to write this up in case you had any further thoughts on this and as a note for others implementing Sentry with Temporal Python SDK.

cretz commented:
Hrmm, this is true. I think we need to be putting contextvars from interceptors on to threaded activities. I have opened https://github.com/temporalio/sdk-python/issues/263.
